hw/ip/lc_ctrl/util/gen-lc-state-enc.py

```diff
@@ -0,0 +1,397 @@
+
+
+def validate(config):
+    '''Validate configuration dict.'''
+
+    if 'secded' not in config:
+        log.error('Missing secded configuration')
+        exit(1)
+
+    config['secded'].setdefault('data_width', 0)
+    config['secded'].setdefault('ecc_width', 0)
+    config['secded'].setdefault('ecc_matrix', [[]])
+    config.setdefault('num_ab_words', 0)
+    config.setdefault('num_cd_words', 0)
+    config.setdefault('num_ef_words', 0)
+    config.setdefault('min_hw', 0)
+    config.setdefault('max_hw', 0)
+    config.setdefault('min_hd', 0)
+
+    config['secded']['data_width'] = _check_int(config['secded']['data_width'])
+    config['secded']['ecc_width'] = _check_int(config['secded']['ecc_width'])
+    config['num_ab_words'] = _check_int(config['num_ab_words'])
+    config['num_cd_words'] = _check_int(config['num_cd_words'])
+    config['num_ef_words'] = _check_int(config['num_ef_words'])
+    config['min_hw'] = _check_int(config['min_hw'])
+    config['max_hw'] = _check_int(config['max_hw'])
+    config['min_hd'] = _check_int(config['min_hd'])
+
+    total_width = config['secded']['data_width'] + config['secded']['ecc_width']
+
+    if config['min_hw'] >= total_width or \
+       config['max_hw'] > total_width or \
+       config['min_hw'] >= config['max_hw']:
+        log.error('Hamming weight constraints are inconsistent.')
+        exit(1)
+
+    if config['max_hw'] - config['min_hw'] + 1 < config['min_hd']:
+        log.error('Hamming distance constraint is inconsistent.')
+        exit(1)
+
+    if config['secded']['ecc_width'] != len(config['secded']['ecc_matrix']):
+        log.error('ECC matrix does not have correct number of rows')
+        exit(1)
+
+    for i, l in enumerate(config['secded']['ecc_matrix']):
+        for j, e in enumerate(l):
+            e = _check_int(e)
+            config['secded']['ecc_matrix'][i][j] = e
+
+
+def _is_valid_codeword(config, codeword):
+    '''Checks whether the bitstring is a valid ECC codeword.'''
+
+    data_width = config['secded']['data_width']
+    ecc_width = config['secded']['ecc_width']
+    if len(codeword) != (data_width + ecc_width):
+        log.error("Invalid codeword length {}".format(len(codeword)))
+        exit(1)
+
+    # Build syndrome and check whether it is zero.
+    syndrome = [0 for k in range(ecc_width)]
+
+    # The bitstring must be formatted as "data bits[N-1:0]" + "ecc bits[M-1:0]".
+    for j, fanin in enumerate(config['secded']['ecc_matrix']):
+        syndrome[j] = int(codeword[ecc_width - 1 - j])
+        for k in fanin:
+            syndrome[j] ^= int(codeword[ecc_width + data_width - 1 - k])
+
+    return sum(syndrome) == 0
+
+
+def _ecc_encode(config, dataword):
+    '''Calculate and prepend ECC bits.'''
+    if len(dataword) != config['secded']['data_width']:
+        log.error("Invalid codeword length {}".format(len(dataword)))
+        exit(1)
+
+    # Build syndrome
+    eccbits = ""
+    for j, fanin in enumerate(config['secded']['ecc_matrix']):
```

Comment:
do you need enumerate here..? looks like j is not being used below.

Reply:
ah yeah that's right. I think I copied that from above... fixed!